source/extensions/filters/http/response_map/response_map_filter.cc

```diff
@@ -0,0 +1,254 @@
+  // is not yet finished, we need to prevent other filters from iterating until
+  // we've had a chance to actually rewrite the body later in encodeData.
+  if (!end_stream) {
+    return Http::FilterHeadersStatus::StopIteration;
+  }
+
+  // Now that the stream is complete, rewrite the response body.
+  ENVOY_LOG(trace, "response map filter: encodeHeaders doing rewrite");
+  doRewrite();
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus ResponseMapFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  // If this filter is disabled, continue without doing anything.
+  if (disabled_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  // If we decided not to rewrite the response, simply pass through to other
+  // filters.
+  if (!do_rewrite_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  // We decided to rewrite the response, so drain any data received from the
+  // upstream since we're rewriting it anyway. This both prevents unnecessary
+  // buffering and also prevents generating errors if the response is too big
+  // to be buffered completely.
+  data.drain(data.length());
+
+  // The stream is not yet complete, and we can't let other filters run
+  // until we've rewritten the response body.
+  if (!end_stream) {
+    return Http::FilterDataStatus::StopIterationAndBuffer;
+  }
+
+  // Now that the stream is complete, rewrite the response body.
+  ENVOY_LOG(trace, "response map filter: encodeData doing rewrite");
+  doRewrite();
+  return Http::FilterDataStatus::Continue;
+}
+
+void ResponseMapFilter::doRewrite(void) {
+  const Buffer::Instance* encoding_buffer = encoder_callbacks_->encodingBuffer();
+
+  ENVOY_LOG(trace, "response map filter: doRewrite with {} encoding_buffer",
+            encoding_buffer != nullptr ? "non-null" : "null");
+
+  // If this route is disabled, we should never be doing a rewrite.
+  // In fact, we never should have even checked if we should do
+  // a rewrite.
+  ASSERT(!disabled_);
+
+  // We should either see no encoding buffer or an empty encoding buffer.
+  //
+  // We'll see no encoding buffer if the upstream response was never observed
+  // (ie: due to error) or if the upstream response was headers-only. Otherwise,
+  // if we did see a response, we should have drained any data we saw.
+  ASSERT(encoding_buffer == nullptr || encoding_buffer->length() == 0);
+
+  // This should be impossible, because we only set response_map_ to the
+  // global filter config (guaranteed to exist if this filter exists)
+  // or to a non-null per-route config. Guard against it anyway.
+  if (response_map_ == nullptr) {
+    ENVOY_LOG(trace,
+              "response map filter: doRewrite has no response_map_ to rewrite with, doing nothing");
+    return;
+  }
+
+  const ResponseMap::ResponseMapPtr& response_map = *response_map_;
+
+  // Fill in new_body and new_content_type using the response map rewrite.
```

Comment:
Extend this comment to mention that we capture the rewritten `code` here per the method signature, but we don't actually do anything with the new value.